### ISSUE TITLE
Benchmarking experiment with AptaNet and AptaTrans

### ISSUE THREAD

**Issue by fkiraly** [gc-os-ai/pyaptamer#109]
Currently, we have two algorithms for binding affinity prediction: AptaNet and AptaTrans, with both producing a classification outcome.

We have also obtained the AptaTrans dataset which can be used for benchmarking. We should try both algorithms in a classical benchmarking experiment, using the train and test splits of the dataset.

* AptaNet also needs to be adapted for regression outcomes, for this
* the experiment would be a jupyter notebook that can be executed, and would lead to classical regression outcomes
* optionally, we need the mfold transformer (tbd)

Edit by @satvshr:
As discussed in #128, we will be using the `AptaCom` dataset to benchamark a regression experiment so the scope of this PR is being increased from being only a classification experiment, to a regression AND classification experiment.

**Comment by satvshr** [comment]
> AptaNet also needs to be adapted for regression outcomes

Can you be a more specific on how you want to adapt it?

**Comment by satvshr** [comment]
> AptaTrans a numerical outcome.

In the AptaTrans PR I [found one dataset](AptaTrans uses a [binary classification dataset](https://github.com/gc-os-ai/pyaptamer/blob/26ba9e694fa727fad1dba762ce919a2d0ca5f9bb/pyaptamer/datasets/data/train_li2014.csv)? I am so confused, I thought it outputted binding affinity?) which uses positive/negative binary targets and not binding affinities, I am a bit confused as to how what dataset to use and also, what is the dataset I linked doing in the AptaTrans PR?

**Comment by fkiraly** [comment]
> I was wondering (also asked this on discord) if you could describe what "benchmarking" entails a bit more as I have never worked on it before.

The basic workflow for benchmarking classifiers or regressors works like this, high-level:

1. apply a splitter or cross-validator, to obtain one or multiple train-test splits on the data (preferably uncorrelated or randomly at uniform, in an iid setting). This is done for each dataset.
2. fit each classifier multiple times on the train split and predict the corresponding test split; collect the predictions
3. carry out a suitable paired/blocked statistical tests using the individual or aggregate predictions from the classifiers. You have pairing/blocking through the dataset and possibly data index. Blocks correspond to classifiers.
4. print and/or plot a report, e.g., of a ranking, or of performances.

The variable parts here are:

* which cross-validation setup to choose
* which data set or data sets to choose
* which classifiers (or regressors) to choose
* how to carry out statistical tests and report the results

For the start, I would suggest:

* simple 50/50 split
* the one data set we have
* the classifiers we have
* Friedman test on paired by-index residuals; critical difference test and critical difference plot

This setup has some limitations (e.g., it does not capture variability arising from fitting), but we can refine and complicate this later.

**Comment by fkiraly** [comment]
> The dataset we have from AptaTrans is already splitted into train-test (original splits from the paper).

I see - in that case I think it is an obvious choice to use precisely those splits.

Of course if you are paranoid, you may want to merge and split uniformly at random and rerun, just to check that there was no "grad student descent" in the original, where repeated use of the same split led to spurious improvements after manual experimentation...

> You will probably need to tune model hyperparameters (e.g., number of layers, learning rate, number of epochs, optimizer, ...). To do so, I would split training data into a train-val split.

In a `scikit-learn` style interface, that is usually *not* done by manually splitting, but by applying something like `GridSearchCV` which does this under the hood, the entire algorithm of splitting and optimizing the hyperparameters is done under the hood!
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

This is one of the main attractions of the sklearn-like interfaces that meta-algorithms simply have the same type as a simple algorithm.

This also raises an important point: we should be able to do tuning with whatever interface we design, the main issue seems to be the string or pdb-like inputs which are not native to sklearn.

> Anyway, one thing to absolutely do is to apply a stratified split w.r.t. the labels, because the dataset is imbalanced in favour of negative labels (no binding). By doing so you will ensure the validation split properly resembles the original data distribution.

Why not. Though it comes pre-split already, no? So this choice does not apply if we use the pre-defined splits.

Regarding "stratified or not": playing the devils advocate, with uniform subsampling, the proportions will *still* resemble the original proportions, if they are not too close to 0/100%, due to the Binomial CLT. Though there are known problems *if* you are too close to 0/100%, see also https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Problems_with_using_a_normal_approximation_or_%22Wald_interval%22, so stratified makes sense.

> For doing these statistical tests, especially the critical difference one, I would suggest the aeon-toolkit library

Let's not use that. It's a copycat project of `sktime`.

**Comment by NennoMP** [comment]
> > The dataset we have from AptaTrans is already splitted into train-test (original splits from the paper).
> 
> I see - in that case I think it is an obvious choice to use precisely those splits.
> 
> Of course if you are paranoid, you may want to merge and split uniformly at random and rerun, just to check that there was no "grad student descent" in the original, where repeated use of the same split led to spurious improvements after manual experimentation...

We could do that.

> Why not. Though it comes pre-split already, no? So this choice does not apply if we use the pre-defined splits.

Stratified splitting was in the context of generating an additional validation split from train data for model selection: validation split for tuning hyperparameters, (unseen) test data for final evaluation.

> Let's not use that. It's a copycat project of `sktime`. The authors of aeon 1:1 copied `sktime` and then shortly after published a paper about it, this is basically plagiarism. We should not support that kind of behaviour, it's academic misconduct.

Ops... Didn't know about that 😁. I used it a few times, guess it's time to find an alternative then.
 
> Why don't we add CD to `sktime` instead?

Could be a cool feature if not already present.

**Comment by fkiraly** [comment]
> Didn't know about that

Feel free to follow up and check - it is all there verifiable from GitHub records. E.g., just check my historical commits to `sktime` vs how they are attributed in the plagiate (not to my GH account because the email address has been intentionally altered). This has the effect of removing me from the zenodo publication record as a formal author, for instance.

**Comment by satvshr** [comment]
> do you need a generic description on how to use it with AptaNet

I need to do that?!

I just wanted to know what needs to bechanged in MCTS for regression 😅

**Comment by NennoMP** [comment]
> > do you need a generic description on how to use it with AptaNet
> 
> I need to do that?!
> 
> I just wanted to know what needs to bechanged in MCTS for regression 😅

Oh, you don't need to change anything in MCTS. It already expects to be fed float values via an experiment.

What you may need to change is something in `AptametEvalAptaNet` from #121. Since it was supposed to work with probabilities scores for binary classes in the `evaluate(...)` method you have `return torch.tensor(score[:, 1])`. If the model is a regressor and gives you float values you would simply return `torch.tensor(score)` I suppose. It depends on how you implement `AptaNet` as a regressor.

Does that make sense? Let me know if I missed something or you have other questions.

**Comment by fkiraly** [comment]
I think what is important to realize that we have different algorithms here:

* prediction algorithms, e.g., regressors or classifiers like AptaNet
* aptamer design algorithms, like MCTS plus AptaNet

I was thinking that the initial benchmarking experiment would be a simple regression or classification experiment.

MCTS does not appear in these, no?
